CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,300 @@
+        void produce(edm::Event&, const edm::EventSetup&) override;
+
+        edm::ESWatcher<HcalRecNumberingRecord> hcalDbWatcher_;
+        edm::ESWatcher<HcalRespCorrsRcd> hcalRCWatcher_;
+  
+        edm::EDGetTokenT<std::vector<reco::PFCandidate> > pfcandidates_;
+
+        std::vector<std::tuple<float,float,float>> badChHE_;
+        std::vector<std::tuple<int,int,int,float>> badChHF_;
+};
+
+PFCandidateRecalibrator::PFCandidateRecalibrator(const edm::ParameterSet &iConfig) :
+    pfcandidates_(consumes<std::vector<reco::PFCandidate>>(iConfig.getParameter<edm::InputTag>("pfcandidates")))
+{
+    produces<std::vector<reco::PFCandidate>>();
+    produces<std::vector<reco::PFCandidate>>("discarded");
+    produces<edm::ValueMap<reco::PFCandidateRef>>();
+}
+
+void PFCandidateRecalibrator::beginRun(const edm::Run &iRun, const edm::EventSetup &iSetup)
+{
+    if (hcalDbWatcher_.check(iSetup) || hcalRCWatcher_.check(iSetup))
+      {
+	//Get Calib Constants from current GT
+	edm::ESHandle<HcalDbService> gtCond;
+	iSetup.get<HcalDbRecord>().get(gtCond);
+	
+	//Get Calib Constants from bugged tag
+	edm::ESHandle<HcalTopology> htopo;
+	iSetup.get<HcalRecNumberingRecord>().get(htopo);
+	const HcalTopology* theHBHETopology = htopo.product();
+	
+	edm::ESHandle<HcalRespCorrs> buggedCond;
+	iSetup.get<HcalRespCorrsRcd>().get("bugged", buggedCond);
+	HcalRespCorrs buggedRespCorrs(*buggedCond.product());
+	buggedRespCorrs.setTopo(theHBHETopology);
+	
+	//access calogeometry
+	edm::ESHandle<CaloGeometry> calogeom;
+	iSetup.get<CaloGeometryRecord>().get(calogeom);
+	const CaloGeometry* cgeo = calogeom.product();
+	const HcalGeometry* hgeom = static_cast<const HcalGeometry*>(cgeo->getSubdetectorGeometry(DetId::Hcal,HcalForward));
+	
+	
+	//fill bad cells HE (use eta, phi)
+	const std::vector<DetId>& cellsHE = hgeom->getValidDetIds(DetId::Detector::Hcal, HcalEndcap);
+	for( auto id : cellsHE)
+	  {
+	    float currentRespCorr = gtCond->getHcalRespCorr(id)->getValue();
+	    float buggedRespCorr = buggedRespCorrs.getValues(id)->getValue();
+	    float ratio = currentRespCorr/buggedRespCorr;
+	    
+	    if(ratio != 1.)
+	      {
+		GlobalPoint pos = hgeom->getPosition(id);
+		badChHE_.push_back(std::make_tuple(pos.eta(),pos.phi(),ratio));
+	      }
+	  }
+	
+	//fill bad cells HF (use ieta, iphi)
+	std::vector<DetId> cellsHF = hgeom->getValidDetIds(DetId::Detector::Hcal, HcalForward);
+	for( auto id : cellsHF)
+	  {
+	    float currentRespCorr = gtCond->getHcalRespCorr(id)->getValue();
+	    float buggedRespCorr = buggedRespCorrs.getValues(id)->getValue();
+	    float ratio = currentRespCorr/buggedRespCorr;
+	    
+	    if(ratio != 1.)
+	      {
+		HcalDetId dummyId(id);
+		badChHF_.push_back(std::make_tuple(dummyId.ieta(), dummyId.iphi(), dummyId.depth(), ratio));
+	      }
+	  }
+      }
+}
+
+void PFCandidateRecalibrator::endRun(const edm::Run &iRun, const edm::EventSetup &iSetup)
+{
+    badChHE_.clear();
+    badChHF_.clear();
```

Comment:
in the current implementation with ESWatchers this reset will lead to a bug: if a new run does not have a change in IOV, then the filling logic will never be called and the badCh vectors will be empty.
It makes more sense to instead clear/reset the values simply right before filling them